File: Object-Oriented-Programming-CPP-2016-/hw8/BigramMap.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Map;
import java.util.HashMap;
import java.util.Scanner;
import java.util.Vector;
import java.lang.Exception;
import java.util.InputMismatchException;
import java.util.Set;
/**
 *
 * @author aktenburakk
 */
@SuppressWarnings("unchecked")
public class BigramMap<T> implements Bigram<T> {
    
    private Map<myPair <T , T> , Integer > pairDatas;//holds pairs ant their occurences.
    private int dataType;//thpe of the data in file.
    private Scanner input ;
    private int max = 0;//holds maximum occurences
    private Vector<myPair<T ,T> > dataPair;//holds data as pair.
     
    private void makePair(Vector<T> data){
    	//this function make all data in vector pair. 
        dataPair = new Vector<myPair<T , T > >();

        for(int i = 0 ; i < data.size() - 1; i++){
            myPair<T,T> temp = new myPair<T , T >(data.get(i),data.get(i + 1));
            dataPair.add(temp);
		}
    }

    private void fillMap(){

    	//this function fill the map with pairs and their occurences.
        myPair<T,T> temp = new myPair<T,T>(dataPair.get(0));
        pairDatas.put(temp , 1);
        max = 1;

        for (int i = 1 ; i < numGrams() ; ++i){
    		int count=1	;
    		temp =(dataPair.get(i));

    		if(pairDatas.containsKey(temp)){
    			count = pairDatas.get(temp);
    			pairDatas.put(temp , count + 1);
    			count++;
    			if(count > max)
    				max = count;
    		}
    		else
    			pairDatas.put(temp , 1);   			
    	}
    }
    
    public BigramMap(int type){//constructor.
        dataType = type;
    }
    
    public void readFile(String fileName)throws Exception {
    	//This function reads file and fill the map.
    	//if any problem happend then throw exception.
    	// for example , empty file , bad data ...
        
        Vector<T> data = new Vector<T> ();//holds the data one by one.
        
        pairDatas =new HashMap<myPair<T ,T> , Integer>();
        
        //dataType is undefined if throw exception here.
        if(dataType <= 0 || dataType > 3){
        	System.err.println("Undefined data thype!!");
        	throw new Exception();
        }

        //reading File/////
        try {

            input = new Scanner(new File(fileName));

            //if file doesn't contain anything.
            if (!input.hasNext()){
        		System.err.println("The file is empty!");
            	throw new Exception();
        	}

        	T next = null ;

        	//this loop read data from the file according to dataType.
        	//the add the data to vector that name is data.
        	while(input.hasNext()){
	        	
		        if(dataType == 1)     
		            (next) = (T)(Object)(input.nextInt());
		        else if (dataType == 2)
	                (next) =  (T)(input.next());
		        else if(dataType == 3)
	                (next) = (T)(Object)(input.nextDouble());
	    		
	            data.addElement(next);
			}
      		makePair(data);//make the data pair.
      		fillMap();//then fill the map. 
        }
        catch (FileNotFoundException ex) {
        	System.err.println("The file is not exist!");
            throw new Exception();
        }
        catch(InputMismatchException e){
	    	System.err.println("Bad Data!!");
	    	throw new Exception();
	    }	
        finally{
        	input.close();
        }
	}
    
    public int numGrams() {//return size of map.
        return dataPair.size();
    }

    public int numOfGrams(T first, T second) {//returns the occurences of the given pair.
        
        int count = 0;
        for(int i = 0 ;i < dataPair.size(); ++i){
        
            if(dataPair.get(i).first.equals(first) && dataPair.get(i).second.equals(second))
                count++;
        }
        
        return count;
    }
   
    public String toString(){
        
    	StringBuilder result = new StringBuilder();//holds the output to print out.
    	Set<myPair<T ,T>> keys = pairDatas.keySet();//holds the keys

    	for(int i = max ; i >= 0 ; --i){
    		for(myPair<T ,T> key : keys){
    			if(pairDatas.get(key) == i)
    				result.append(key + "->" + pairDatas.get(key) + "\n");
    		}
    	}
        return result.toString();
    }
}

File: Object-Oriented-Programming-CPP-2016-/hw7/BigramMap.cpp
/* 
 * File:   BigramMap.cpp
 * Author: aktenburakk
 * 
 * Created on 10 December 2016, 11:45
 */

#include "BigramMap.h"
#include "Bigram.h"
#include <string>
#include <iostream>
#include <map>
#include <fstream>
using namespace std;

template <class T>
BigramMap<T>::BigramMap(int type) : Bigram<T>() , dataType(type){}

template <class T>
void BigramMap<T>::readFile(string fileName) throw(Error){

    ifstream inp;
    
    vector<T> datas;//keep all datas one by one.
    
    inp.open(fileName);//file is opening
    if(inp.fail()){//if file is not exist then throwing Error object
    	cerr << "No such a file!!"<<endl;
        throw Error();
    }

    T inpData;
    //reads the file until end of file
    while (!inp.eof()  && inp >> inpData)
        datas.push_back(inpData);
    
    if(!inp.eof()){
        cerr<<"Bad Data!"<<endl;
        inp.close();
        throw Error();
    }
    inp.close();

    //if file is empty then throw Error exception.
    if(datas.size() == 0){
        cerr<<"Empty File!!!"<<endl;
        throw Error();
    }
    
    //the data that read from file is being made pairs by this function
    makePair(datas);
    //fill the map is numberOfDatas with these pair and their occurences.
    fillMap();
}

template <class T>
void BigramMap<T>:: print()const{
    //reverse iterator for beginning form the end of the map to printo out.
    typename map<pair<T , T> , int >::const_reverse_iterator p ;
    typename map<pair<T , T> , int >::const_reverse_iterator temp ;
    //first loop begins with max. occurences and goes on until zero.
    //to print out the all occurences of all pairs.
    for(int i = numberOfMaxGrams; i > 0 ; --i)
        for(p = numberOfDatas.rbegin();p != numberOfDatas.rend() ; p++){
            if(p->second == i){         
                cout<<"("<<p->first.first<<"-"<< p->first.second<<") =>";
                cout << p->second<<endl; 
            }
        }
}

template<class T>
void BigramMap<T>:: fillMap(){

	//fill the map with pairs and their occurrences.
    for(int i = 0; i < pairDatas.size() ; ++i){
        pair<T , T> temp = pairDatas[i];
        //calculates occurence of that pair.
        int count = numOfGrams(temp.first , temp.second);

        numberOfDatas[pairDatas[i]] = count;
    }

}

template <class T>
void BigramMap<T>::makePair(std::vector<T>& vec){
    //it makes all data pair into pairDatas.
    for(int i = 0; i < vec.size() - 1 ; ++i)
        pairDatas.push_back(make_pair(vec[i] , vec[i + 1]));
}

template <class T>
int BigramMap<T>:: numGrams()const {return pairDatas.size();}

template <class T>
int BigramMap<T>::numOfGrams(T first , T second)const{

	//holds occurences
    int total = 0;
    //make the paramaters pair.
    pair<T ,T> temp = make_pair(first , second);
    
    for(int i = 0 ; i < numGrams() ; ++i){
        if(pairDatas[i] == temp)
            total++;
    }
    return total;
}

template <class T>
pair<T ,T> BigramMap<T>:: maxGrams(){

    int max = 0;
    
    typename map<pair<T , T>, int >::const_iterator p ;

    //This first loop find the maximum occurence in fale as a pair.
    for(p = numberOfDatas.begin();p != numberOfDatas.end() ; p++)
        if(p->second > max)
            max = p->second;

    pair<T ,T> temp;
    bool found = false;
    //This part firstly find out which pairs occurences equal to max and if 
    //this pair is occurence in the file as first one then it is returned.
    for (int i = 0; i < pairDatas.size() && !found; ++i)
    	for(p = numberOfDatas.begin(); p != numberOfDatas.end() && !found ; p++)
    		if (p->first == pairDatas[i] && p->second == max){
    			temp = make_pair(p->first.first , p->first.second);
    			found = true;
    		}
    
    numberOfMaxGrams = max;//initialize the value of numberOfMaxGrams with max.

    return temp;
}



